### ISSUE TITLE
Literal types get turned into basic types (ie string) during evaluation of union types.

### ISSUE THREAD

**Issue by afrieder** [Microsoft/TypeScript#23643]
**TypeScript Version:**  2.8.3

**Search Terms:** 
unions, literal types

**Code**
```ts
type A = {
    a: any,
    b: number,
};

type B = {
    a: { hi: 'hi' },
    b: boolean,
};

const y: A | B = {
    a: { hi: 'hi' },
    b: true,
};
```

**Expected behavior:**
This type-checks since `y` is an instance of `B`.

**Actual behavior:**
This raises an error that `y` is not assignable to `B` because `{ hi: string }` is not assignable to `{ hi: 'hi' }`.

**Playground Link:** 
[Right here](http://www.typescriptlang.org/play/#src=type%20A%20%3D%20%7B%0D%0A%20%20%20%20a%3A%20any%2C%0D%0A%20%20%20%20b%3A%20number%2C%0D%0A%7D%3B%0D%0A%0D%0Atype%20B%20%3D%20%7B%0D%0A%20%20%20%20a%3A%20%7B%20hi%3A%20'hi'%20%7D%2C%0D%0A%20%20%20%20b%3A%20boolean%2C%0D%0A%7D%3B%0D%0A%0D%0Aconst%20y%3A%20A%20%7C%20B%20%3D%20%7B%0D%0A%20%20%20%20a%3A%20%7B%20hi%3A%20'hi'%20%7D%2C%0D%0A%20%20%20%20b%3A%20true%2C%0D%0A%7D%3B)

**Related Issues:**

**Comment by afrieder** [comment]
@sandersn

**Comment by mhegazy** [comment]
the literal type is only inferred if the the contextual type has a matching literal type. in this case the contextual type is `any` | { hi: "hi" }`; `any` subsumes any type really, so the union is reduced to `any` and thus the literal type is not picked up.
This is a design limitation of how union types and contextual type interact, It is worth nothing that any type that is not `any` (e.g. `{} | null | undefined`) would work here. I do not think we should carry `any`'s in unions just to enable this scenario, seems like too much perf hit for all use cases just to avoid this corner case.
